test/server/config_validation/xds_fuzz.cc

```diff
@@ -0,0 +1,244 @@
+  } else {
+    ip_version_ = Network::Address::IpVersion::v6;
+  }
+
+  if (input.config().client_type() == test::server::config_validation::Config::GOOGLE_GRPC) {
+    client_type_ = Grpc::ClientType::GoogleGrpc;
+  } else {
+    client_type_ = Grpc::ClientType::EnvoyGrpc;
+  }
+
+  if (input.config().sotw_or_delta() == test::server::config_validation::Config::SOTW) {
+    sotw_or_delta_ = Grpc::SotwOrDelta::Sotw;
+  } else {
+    sotw_or_delta_ = Grpc::SotwOrDelta::Delta;
+  }
+}
+
+/**
+ * initialize an envoy configured with a fully dynamic bootstrap with ADS over gRPC
+ */
+void XdsFuzzTest::initialize() {
+  config_helper_.addConfigModifier([this](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    auto* ads_config = bootstrap.mutable_dynamic_resources()->mutable_ads_config();
+    auto* grpc_service = ads_config->add_grpc_services();
+
+    std::string cluster_name = "ads_cluster";
+    switch (client_type_) {
+    case Grpc::ClientType::EnvoyGrpc:
+      grpc_service->mutable_envoy_grpc()->set_cluster_name(cluster_name);
+      break;
+    case Grpc::ClientType::GoogleGrpc: {
+      auto* google_grpc = grpc_service->mutable_google_grpc();
+      google_grpc->set_target_uri(xds_upstream_->localAddress()->asString());
+      google_grpc->set_stat_prefix(cluster_name);
+      break;
+    }
+    default:
+      NOT_REACHED_GCOVR_EXCL_LINE;
+    }
+    auto* ads_cluster = bootstrap.mutable_static_resources()->add_clusters();
+    ads_cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+    ads_cluster->set_name("ads_cluster");
+  });
+  setUpstreamProtocol(FakeHttpConnection::Type::HTTP2);
+  HttpIntegrationTest::initialize();
+  if (xds_stream_ == nullptr) {
+    createXdsConnection();
+    AssertionResult result = xds_connection_->waitForNewStream(*dispatcher_, xds_stream_);
+    RELEASE_ASSERT(result, result.message());
+    xds_stream_->startGrpcStream();
+  }
+}
+
+void XdsFuzzTest::close() {
+  cleanUpXdsConnection();
+  test_server_.reset();
+  fake_upstreams_.clear();
+}
+
+/**
+ * remove a listener from the list of listeners if it exists
+ * @param the listener number to be removed
+ * @return the listener as an optional so that it can be used in a delta request
+ */
+absl::optional<std::string> XdsFuzzTest::removeListener(uint32_t listener_num) {
+  std::string match = absl::StrCat("listener_", listener_num % num_listeners_);
+
+  for (auto it = listeners_.begin(); it != listeners_.end(); ++it) {
+    if (it->name() == match) {
+      listeners_.erase(it);
+      return match;
+    }
+  }
+  return {};
+}
+
+/**
+ * remove a route from the list of routes if it exists
+ * @param the route number to be removed
+ * @return the route as an optional so that it can be used in a delta request
+ */
+absl::optional<std::string> XdsFuzzTest::removeRoute(uint32_t route_num) {
+  std::string match = absl::StrCat("route_config_", route_num % num_routes_);
+  for (auto it = routes_.begin(); it != routes_.end(); ++it) {
```

Comment:
Why not make these `absl::flat_hash_map`?

Reply:
good idea - will do

Reply:
Don't feel super strongly, but it could be a good excuse to explore the STL algorithm llibrary, e.g. would `std::remove_if` or something like that work?

Reply:
I thought about using `std::erase(std::remove(...))` but since I return the name of the route that was removed, I needed to check if I had actually removed anything.  If I was to use `std::erase` then I'd probably have to compare the sizes of the vector before and after the erase to find this out so I thought this logic might have been a bit clearer (also since I'm only ever removing one element), but it's up to you, I can definitely change it if you think that approach is better.

Reply:
Sure, it's fine, whatever is clearest for test code makes sense.